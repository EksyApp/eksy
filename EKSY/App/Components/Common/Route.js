import React, {Component} from 'react'
import MapView from 'react-native-maps'
import PropTypes from 'prop-types';
import {MarkersShape, RegionShape, StyleShape} from '../../Utils/PropTypeShapes'

export default class Route extends Component {
	
	
	render() {
		return (
				<MapView
						style={this.props.style}
						initialRegion={{
							...this.props.initialRegion,
							latitude: this.props.markers[0].latitude,
							longitude: this.props.markers[0].longitude
						}}
				>
					{this.props.markers.map((marker) => {
						return(
								<MapView.Marker coordinate={{latitude: marker.latitude, longitude: marker.longitude}}/>
						)
					})}
					
					<MapView.Polyline
							coordinates={this.props.markers.map((marker) => {
								return {latitude: marker.latitude, longitude: marker.longitude}
							})}
					/>
				</MapView>
		)
	}
}

Route.propTypes = {
	style: StyleShape ,
	initialRegion: RegionShape,
	markers: MarkersShape
}