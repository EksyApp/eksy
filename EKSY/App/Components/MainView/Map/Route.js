import React, {Component} from 'react'
import {View} from 'react-native'
import Markers from './Markers'
import Geometry from '../../../Utils/Geometry'
import Marker from './Marker'
import MapView from 'react-native-maps'
import PropTypes from 'prop-types'
import {LocationShape, MarkerShape, RouteShape} from '../../../Utils/PropTypeShapes'

export default class Route extends Component {
	
	constructor(props) {
		super(props)
		this.nextMarkerKey = null
		
		this.state = {
			previousMarkers: this.getPreviousMarkers(this.props.nextMarker)
		}
	}
	
	getPreviousMarkers(nextMarker) {
		let previousMarkers = []
		for (let marker of this.props.route.markers) {
			if (marker.key !== nextMarker.key) {
				previousMarkers.push(marker)
			} else {
				break
			}
		}
		return previousMarkers
	}
	
	componentWillReceiveProps(props) {
		if (props.nextMarker.key !== this.nextMarkerKey) {
			this.nextMarkerKey = props.nextMarker.key
			this.setState({previousMarkers: this.getPreviousMarkers(props.nextMarker)})
		}
		
	}
	
	renderNextMarker() {
		if (Geometry.distance(this.props.currentLocation, this.props.nextMarker) >= this.props.radius) {
			return (
					<Marker
							marker={{
								...Geometry.getPointOnRadius(this.props.currentLocation, this.props.nextMarker, this.props.radius),
								status: 2
							}}
							onPress={() => {}}
					/>
			)
		} else {
			return (
					<Marker
							marker={this.props.nextMarker}
							onPress={(marker) => this.props.onMarkerClick(marker, this.state.previousMarkers.length)}
					/>
			)
		}
	}
	
	renderPointerToNextMarker() {
		if(this.state.previousMarkers.length > 0) {
			let lastVisited = this.state.previousMarkers[this.state.previousMarkers.length-1]
			let pointerLength = 0.1
			if(Geometry.distance(lastVisited, this.props.nextMarker) >= pointerLength) {
				return <MapView.Polyline coordinates={[lastVisited, Geometry.getPointOnRadius(lastVisited, this.props.nextMarker, pointerLength)]}/>
			} else {
				return <MapView.Polyline coordinates={[lastVisited, this.props.nextMarker]}/>
			}
		}
		
	}
	
	render() {
		return (
				<View>
					<Markers
							markerList={this.state.previousMarkers}
							onMarkerClick={this.props.onMarkerClick}
					/>
					{this.renderNextMarker()}
					<MapView.Polyline coordinates={this.state.previousMarkers}/>
					{this.renderPointerToNextMarker()}
				</View>
		)
	}
	
	
	
}

Route.propTypes = {
	route: RouteShape,
	nextMarker: MarkerShape,
	currentLocation: LocationShape,
	radius: PropTypes.number,
	onMarkerClick: PropTypes.func
}