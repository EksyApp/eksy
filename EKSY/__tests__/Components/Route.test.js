import React from 'react';
import renderer from 'react-test-renderer';
import Route from "../../App/Components/Common/RouteMap";

jest.mock('react-native-maps', () => {
	const React = require.requireActual('react');
	const MapView = require.requireActual('react-native-maps');
	
	class MockCallout extends React.Component {
		render() {
			return React.createElement('Callout', this.props, this.props.children);
		}
	}
	
	class MockMarker extends React.Component {
		render() {
			return React.createElement('Marker', this.props, this.props.children);
		}
	}
	
	class MockMapView extends React.Component {
		render() {
			return React.createElement('MapView', this.props, this.props.children);
		}
	}
	
	class MockMapViewAnimated extends React.Component {
		render() {
			return React.createElement('MapView', this.props, this.props.children);
		}
	}
	
	class MockPolyline extends React.Component {
		render() {
			return React.createElement('Polyline', this.props, this.props.children);
		}
	}
	
	MockCallout.propTypes = MapView.Callout.propTypes;
	MockMarker.propTypes = MapView.Marker.propTypes;
	MockMapView.propTypes = MapView.propTypes;
	MockMapView.Marker = MockMarker;
	MockMapView.Callout = MockCallout;
	MockMapView.Animated = MockMapViewAnimated;
	MockMapView.Polyline = MockPolyline;
	return MockMapView;
});

test('RouteMap renders correctly', () => {
	const tree = renderer.create(
			<Route
					initialRegion = {{latitude: 1, longitude: 1, latitudeDelta: 1, longitudeDelta: 1}}
					markers = {[{latitude: 1, longitude: 1}]}
			/>
	).toJSON();
	expect(tree).toMatchSnapshot();
});