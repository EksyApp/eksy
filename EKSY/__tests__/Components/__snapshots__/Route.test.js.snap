// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route renders correctly 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 1,
      "latitudeDelta": 1,
      "longitude": 1,
      "longitudeDelta": 1,
    }
  }
  style={undefined}
>
  <Marker
    coordinate={
      Object {
        "latitude": 1,
        "longitude": 1,
      }
    }
  />
  <Polyline
    coordinates={
      Array [
        Object {
          "latitude": 1,
          "longitude": 1,
        },
      ]
    }
  />
</MapView>
`;
